### ISSUE TITLE
MySensors TempHumBaro

### ISSUE THREAD

**Issue by nekitoss** [domoticz/domoticz#2902]
@gizmocuz 
While debugging [issue#2878](https://github.com/domoticz/domoticz/issues/2878) i think i've found bug related to temp+humidity+baro combination of childs at one node.
My version is `v4.9700` for RPi and `v4.10240` for linux gdb debugging

The **main problem is**: depending on the order of sending data from sensors (temp/hum/pressure) one of them will not appear (not even in combined form). And sensor that not appear is the last one in order. All corresponding childs are present and have thair values.

First i tried it as described [on forum](https://forum.mysensors.org/topic/9924/bme280-sensor-missing-combining-domoticz-heartbeat-and-battery-level).
Then i started debugging with gdb and found that function `MySensorsBase::SendSensor2Domoticz` works perfectly, but when third sensor appears it (depending on order) get to line [637](https://github.com/domoticz/domoticz/blob/master/hardware/MySensorsBase.cpp#L637) or [720](https://github.com/domoticz/domoticz/blob/master/hardware/MySensorsBase.cpp#L720) or [795](https://github.com/domoticz/domoticz/blob/master/hardware/MySensorsBase.cpp#L795) which are the same (second parameter just have different variable name, but always taken from temperature). Problem starts/hides in that line(s).
Also i tried all possible combinations (results on forum and one case on screens below) and every time i was deleting devices and hardware and restarting domoticz `sudo service domoticz.sh restart`.
No errors appeared in http-domoticz-setup-log

My very simplified code looks like:
```
#define PRES_BME_ID 0
#define TEMP_BME_ID 1
#define HUMID_BME_ID 2


MyMessage pressure_msg(PRES_BME_ID, V_PRESSURE);
MyMessage temperature_msg(TEMP_BME_ID, V_TEMP);
MyMessage humidity_msg(HUMID_BME_ID, V_HUM);


void presentation()
{
  sendSketchInfo("lodjiya", "0.1");

  present(PRES_BME_ID, S_BARO, "lodjiya_bme_pres");
  present(TEMP_BME_ID, S_TEMP, "lodjiya_bme_t");
  present(HUMID_BME_ID, S_HUM, "lodjiya_bme_humid");
}

void loop()
{
  send(pressure_msg.set(p));
  send(temperature_msg.set(t, 1));
  send(humidity_msg.set(h));
}
```

The **second problem** i think raise from first: existing 2 of 3 _devices_ are "stucked" and will not update data, while corresponding _childs_ in _hardware_ are updated normally and have latest values.

Devices:
![stucked_device](https://user-images.githubusercontent.com/11498479/49565122-cbee3d00-f92e-11e8-8d9f-e5b9b5b23b52.png)

Hardware:
![not_stucked_hardware](https://user-images.githubusercontent.com/11498479/49565049-7f0a6680-f92e-11e8-9988-328d247fe5d1.png)

As you can see - in childs are 3 sensors (t/h/p) but as for code first was send&created pressure, then temperature, and then it tried to create device "TempHumBaro" (according to code), but it didn't...
Also previously created _devices_ hadn't updated their values, while _childs_ already are (look at red  rectangular).
P.S. do not look at missing data for lux - because it was too dark it will not send data

**Comment by gizmocuz** [comment]
Please copy/paste your images directly into github, i can't see them now
I also changed the topic of this issue to include MySensors
You need to give am all the same ID as they belong to the same unit

**Comment by nekitoss** [comment]
About images - strange, i drag&drop them into editor window. Also checked in incognito window - i still can see images... Maybe you and git-img-server had some connection problems...
Or you wanted me to paste from clipboard? (this post i ctrl+v image in editor - they appear same way as previous....)

Now i gave these three sensors same id. (And added send battery, so you can see where it is)
my code now looks like:
```
#define PRES_BME_ID 0
#define TEMP_BME_ID 0
#define HUMID_BME_ID 0


MyMessage pressure_msg(PRES_BME_ID, V_PRESSURE);
MyMessage temperature_msg(TEMP_BME_ID, V_TEMP);
MyMessage humidity_msg(HUMID_BME_ID, V_HUM);


void presentation()
{
  sendSketchInfo("lodjiya", "0.1");

  present(PRES_BME_ID, S_BARO, "lodjiya_bme_pres");
  present(TEMP_BME_ID, S_TEMP, "lodjiya_bme_t");
  present(HUMID_BME_ID, S_HUM, "lodjiya_bme_humid");
}

void loop()
{
  sendBatteryLevel(batteryPcnt);
  send(pressure_msg.set(p));
  send(temperature_msg.set(t, 1));
  send(humidity_msg.set(h));
}
```

then i deleted old devices, then deleted whole node from hardware and restarted domoticz (at this time firmware was uploading to node, so node wasn't sending sensor data or presenting)

And... the same thing
hardware:
![image](https://user-images.githubusercontent.com/11498479/49635051-28bd2680-fa07-11e8-84d1-feaa7d7f5c55.png)
devices:
![image](https://user-images.githubusercontent.com/11498479/49635063-3672ac00-fa07-11e8-93df-4bb55a9e62e5.png)
still no errors appear in web-log
still stucked data
still 2 of 3

also noticed that as for my question about battery level (and heartbeat) in other issue it updated battery level (which is send before sensors data) only for those sensors, who send their data.

This test was only on `v4.9700`

**Comment by gizmocuz** [comment]
i think it is better to use '1' as ID, not 0...

I don't see the humidity value ? Is accept new devices enabled ? If not, do some for some time
Yes you can get a separate temp hum or baro sensor, but when all three are received then it should update a combined one

**Comment by nekitoss** [comment]
Does Domoticz doesn't like '0' id? I used zero just because childs numeration starts from zero

Yes, that is the problem - humidity value is present and updated in **hardware**: `V_HUM (33)`, _but not in devices_. As i pointed earlier - it is because humidity is who last sends. If i send last, for example, pressure - then combined humidity and temperature device will be created, but pressure won't be present **in devices**. 
If i (start from scratch) send last temperature - then pressure and humidity devices will be created, but pressure won't be present in devices.
 (I'm sending float for temerature and integers for humidity and pressure - don't think that must be a problem...)

But if you are saying that all 3 must be created **first and separately**, so combined sensor  then _to be possible to be created_(?) and updated then maybe code is wrong... Because looks like devices are created when they send their first data, aren't they? (Possibly except for mentioned in lines [2256](https://github.com/domoticz/domoticz/blob/master/hardware/MySensorsBase.cpp#L2256); [2273](https://github.com/domoticz/domoticz/blob/master/hardware/MySensorsBase.cpp#L2273); [2281](https://github.com/domoticz/domoticz/blob/master/hardware/MySensorsBase.cpp#L2281): `V_STATUS, V_RGB, V_RGBW, V_TRIPPED, V_PERCENTAGE, V_LOCK_STATUS, V_UP, V_TEXT`)

Then according to code for cases V_TEMP-VHUM-V_PRESSURE [580](https://github.com/domoticz/domoticz/blob/master/hardware/MySensorsBase.cpp#L580)-[654](https://github.com/domoticz/domoticz/blob/master/hardware/MySensorsBase.cpp#L654)-[737](https://github.com/domoticz/domoticz/blob/master/hardware/MySensorsBase.cpp#L737) any of them **first** will look if other two exist and if yes - send combined, else send separated.
So for order of sending data p-t-h:
1. P: no other exist, send separated pressure -> here pressure device is created?
2. H: only P exist, so send separated humidity -> here humidity device is created?
3. T: bot P and H exist, so sen combined of 3 `TempHumBaro` sensor -> no separate humidity will be created, but instead must be combined created, but it can't because separate humidity doesn't exists?

Note, that i've looked and debugged only MysensorsBase, so in case if sending triplecombined sensor will create separated missing sensors or they are created somewhere else - then i'm fully wrong...

**Comment by gizmocuz** [comment]
@nekitoss , sure you must also think that 0 is not an ID ?

Yes debugging MySensorsBase is a good idea.

Make sure accept new devices is enabled in the settings

**Comment by nekitoss** [comment]
@gizmocuz about id - i just don't understood idea why it is better 1 than 0 =) 
More readable and habitual?
"i debugged only MysensorsBase" - so i don't know what is happening in other parts, it would be overload for me...
![image](https://user-images.githubusercontent.com/11498479/49671813-c3eae600-fa71-11e8-9d2d-f2b72a45f486.png)

Are _devices_ created when they send teir first data?

**Comment by gizmocuz** [comment]
If you set a breakpoint where the temp/hum/baro sensors are being send, you should be able to debug the issue

**Comment by gizmocuz** [comment]
Found anything during debugging ?

**Comment by nekitoss** [comment]
Yes, sorry @gizmocuz , i was busy.
It took me a lot of time... because i don't know how debug in threads
So problem was in my side - i didn't know, that domoticz require pressure in hectopascals (hPa) (same as millibars (mbar)). I was sending data in our usual millimeters of mercury (mmHg)...
So at mainworker.cpp decoding was stopping at [line 3879](https://github.com/domoticz/domoticz/blob/development/main/mainworker.cpp#L3879), because value in mmHg won't fit that numbers..
But for case when baro is not last of them - it will send as only baro sensor - which is implemented as general device at the moment - that why it was creating baro sensor (if not last in order) but not creating tripple sensor...
I did a little search at wiki - didn't find that it must be in hPa..
Also in settings->meters/counters there is no option to choose units for barometer... If we are not only who use something other then hPa it would be great to implement this.
So now, i divide my sensors Pascal result with 100 and send data - all ok, tripple sensor is created.
Same problem people will have if they have humidity > 100 or temerature < -200 or >380...
(But what if somebody really measuring something hot at home? yes, he can use text then...)
The other thing was that there were nothing about wrong pressure in errors logs tab in web interface. But i see that it must write error to log at [line 3881](https://github.com/domoticz/domoticz/blob/development/main/mainworker.cpp#L3881)...

Where do i check that log next time or how to enable debug logs for this - so i wouldn't waste your time? Also couldn't find about this in wiki... 
I found only debug option for dummy devices in hardware...
